### PR TITLE
Replace / by OR in the license field of Rust crates

### PR DESCRIPTION
**Pull request by gferon**
`cargo` supports arbitrary formats to specify multiple licenses, but `/` should be interpreted as `OR`.

It looks like the consensus is that crates.io should [only accept SPDX 2.1 license expressions](https://doc.rust-lang.org/cargo/reference/manifest.html#the-license-and-license-file-fields), but this is not the reality for all crates as it's not validated.
  
Related issue in cargo: https://github.com/rust-lang/cargo/issues/2039

### CODE REVIEW

**Comment by sschuberth** [comment]
Thanks for this @gferon, but can you please put the text from your comment in the commit message body?

**Comment by gferon** [comment]
> Thanks for this @gferon, but can you please put the text from your comment in the commit message body?

instead of in the code or both?

**Comment by gferon** [comment]
> Thanks for this @gferon, but can you please put the text from your comment in the commit message body?

done. it looks like there's an unrelated test failure in CI (in the Sbt analyzer).

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 1]
```diff
@@ -90,10 +90,9 @@ class Cargo(
         VcsHost.toVcsInfo(extractRepositoryUrl(node))
 
     private fun extractDeclaredLicenses(node: JsonNode) =
-        node["license"].textValueOrEmpty().split("/")
-            .map { it.trim() }
-            .filter { it.isNotEmpty() }
-            .toSortedSet()
+        sortedSetOf<String>().apply {
+            add(node["license"].textValueOrEmpty().replace("/", " OR ").trim())
+        }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:95]: This could theoretically be written as `sortedSetOf(node["license"].textValueOrEmpty().replace("/", " OR ").trim())`, but I actually liked the trimming of the individual elements in the original code.

Anyway, I believe we have a more fundamental API problem here: The return value of this function is used to populate the `declaredLicenses` field, and that's defined as `val declaredLicenses: SortedSet<String>`. Strictly speaking we somewhat the same problem with the individual entries in the set: It's unclear whether set members have an `AND` or `OR` relation.

IIRC we deliberately did not *want* to chose *any* relation here, but `declaredLicenses` should literally take whatever string is used in the package manager's meta data. And as some package managers have multiple license fields, `declaredLicenses` is a collection / set instead of a single string.

Long story short, I'm not sure if even the original split by `/` was correct, as that's already not literally taking the meta data. I believe this needs some discussion.

Comment (gferon) [reply analyzer/src/main/kotlin/managers/Cargo.kt:95]: Well, what is there to discuss? I just read the whole story and in the Rust ecosystem the `/` was used as a `OR` in a SPDX license expression, which is what my commit tries to address.

If you believe the toolkit should get the raw value, and it can handle it, I can change it so we don't process anything here. In the meantime, I'd argue that it's better to cover more cases (correctly) over anything else.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Cargo.kt:95]: > It's unclear whether set entries have an `AND` or `OR` relation among each other.

Actually, looking again at how we generate `declaredLicensesProcessed`, I realize [we *do* relate set entries always with `AND`](https://github.com/oss-review-toolkit/ort/blob/376555874ab7e46a95e894a2702e15b94326ee85/model/src/main/kotlin/Package.kt#L60-L64).

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Cargo.kt:95]: To conclude my thinking-out-aloud: I'd be fine with the intent of this change, but I'd implement it as

    private fun extractDeclaredLicenses(node: JsonNode) =
        sortedSetOf(
            node["license"].textValueOrEmpty().split('/')
                .map { it.trim() }
                .filter { it.isNotEmpty() }
                .joinToString(" OR ")
        )

Let's see what others say.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Cargo.kt:95]: Basically I agree with interpreting the `/` as `or`.
I wonder how the `and` operator looks like in Rust - any idea?

Comment (gferon) [reply analyzer/src/main/kotlin/managers/Cargo.kt:95]: @fviernau it doesn't exist, which is why the recommendation was to start using SPDX expressions instead.

Comment (gferon) [reply analyzer/src/main/kotlin/managers/Cargo.kt]: @sschuberth I have applied your suggestion.

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
Looks like we have [test failures](https://ci.appveyor.com/project/oss-review-toolkit/ort/builds/34880217/tests) as probably expected results now need to be updated. Could you please amend you commit with the necessary changes, @gferon?

**Comment by sschuberth** [comment]
Thanks for the update @gferon, but please squash your commits as we aim for each commit in a PR to pass tests on its own.

**Comment by gferon** [comment]
> Thanks for the update @gferon, but please squash your commits as we aim for each commit in a PR to pass tests on its own.

why not switch the contribution model to squash on merge then?

**Comment by sschuberth** [comment]
> why not switch the contribution model to squash on merge then?

Because we *do* want to keep individual commits in PRs and the final history (as long as they pass tests individually) for better / more fine grained traceability of the development process, and to improve running tools like `git bisect`.

**Comment by sschuberth** [comment]
> EDIT: done

Well, you of course need to adjust your commit message then to not contain two sign-offs... that is, just remove all of the second's commit message, and stick with only the message of the first commit.

And while at it, please also prefix the commit message with "Cargo: " to highlight the scope of modified files.

**Comment by gferon** [comment]
@sschuberth I just updated the test files, they should pass now. I also had to change the code a little, feel free to suggest something else, I don't really know Kotlin.

File: analyzer/src/main/kotlin/managers/Cargo.kt [hunk 2]
```diff
@@ -89,11 +90,13 @@ class Cargo(
...
 
-    private fun extractDeclaredLicenses(node: JsonNode) =
-        node["license"].textValueOrEmpty().split("/")
+    private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
+        val licenses = node["license"].textValueOrEmpty().split('/')
             .map { it.trim() }
             .filter { it.isNotEmpty() }
-            .toSortedSet()
+
+        return if(licenses.isEmpty()) sortedSetOf<String>() else sortedSetOf(licenses.joinToString(" OR "))
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Cargo.kt:98]: This should be (note the space after `if`, and the removed explicit type):

    if (licenses.isEmpty()) sortedSetOf() else sortedSetOf(licenses.joinToString(" OR "))

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Cargo.kt (hunk 2 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Cargo.kt:98]: Is the case distinction actually needed, e.g. why not just `sortedSetOf(licenses.joinToString(" OR "))`?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Cargo.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Cargo.kt:98]: Because that would create a `SortedSet` containing an empty string if `licenses` is empty, and not the wanted empty set.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Cargo.kt:98]: Oh right, fully missed that.

Comment (gferon) [reply analyzer/src/main/kotlin/managers/Cargo.kt:98]: > `if (licenses.isEmpty()) sortedSetOf() else sortedSetOf(licenses.joinToString(" OR "))`

When I tried that, it didn't compile, I can try again, though.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Cargo.kt]: I also just tried, for me it compiles fine without the explicit type.

Comment (gferon) [reply analyzer/src/main/kotlin/managers/Cargo.kt]: done.

**Comment by gferon** [comment]
took a while, but tests are finally passing :balloon:

**Review by sschuberth — APPROVED** [review]
